book-content/chapters/10-deriving-types.md

```diff
@@ -0,0 +1,1460 @@
+### `Exclude`
+
+The `Exclude` utility type is used to remove types from a union. Let's imagine that we have a union of different states our album can be in:
+
+```typescript
+type AlbumState =
+  | {
+      type: "released";
+      releaseDate: string;
+    }
+  | {
+      type: "recording";
+      studio: string;
+    }
+  | {
+      type: "mixing";
+      engineer: string;
+    };
+```
+
+We want to create a type that represents the states that are not "released". We can use the `Exclude` utility type to achieve this:
+
+```ts twoslash
+type AlbumState =
+  | {
+      type: "released";
+      releaseDate: string;
+    }
+  | {
+      type: "recording";
+      studio: string;
+    }
+  | {
+      type: "mixing";
+      engineer: string;
+    };
+
+// ---cut---
+type UnreleasedState = Exclude<AlbumState, { type: "released" }>;
+//   ^?
+```
+
+In this case, the `UnreleasedState` type is a union of the `recording` and `mixing` states, which are the states that are not "released". `Exclude` filters out any member of the union with a `type` of `released`.
+
+We could have done it by checking for a `releaseDate` property instead:
+
+```typescript
+type UnreleasedState = Exclude<AlbumState, { releaseDate: string }>;
+```
+
+This is because `Exclude` works by pattern matching. It will remove any type from the union that matches the pattern you provide.
+
+This means we can use it to remove all strings from a union:
+
+```ts twoslash
+type Example = "a" | "b" | 1 | 2;
+
+type Numbers = Exclude<Example, string>;
+//   ^?
+```
+
+### `NonNullable`
+
+`NonNullable` is used to remove `null` and `undefined` from a type. This can be useful when extracting a type from a partial object:
+
+```ts twoslash
+type Album = {
+  artist?: {
+    name: string;
+  };
+};
+
+type Artist = NonNullable<Album["artist"]>;
+//   ^?
+```
+
+This operates similarly to `Exclude`:
+
+```typescript
+type Artist = Exclude<Album["artist"], null | undefined>;
+```
+
+But `NonNullable` is more explicit and easier to read.
+
+### `Extract`
+
+`Extract` is the opposite of `Exclude`. It's used to extract types from a union. For example, we can use `Extract` to extract the `recording` state from the `AlbumState` type:
+
+```ts twoslash
+type AlbumState =
+  | {
+      type: "released";
+      releaseDate: string;
+    }
+  | {
+      type: "recording";
+      studio: string;
+    }
+  | {
+      type: "mixing";
+      engineer: string;
+    };
+
+// ---cut---
+type RecordingState = Extract<AlbumState, { type: "recording" }>;
+//   ^?
+```
+
+This is useful when you want to extract a specific type from a union you don't control.
+
+Similarly to `Exclude`, `Extract` works by pattern matching. It will extract any type from the union that matches the pattern you provide.
+
+This means that, to reverse our `Extract` example earlier, we can use it to extract all strings from a union:
+
+```typescript
+type Example = "a" | "b" | 1 | 2 | true | false;
+
+type Strings = Extract<Example, string>;
```

Comment:
a powerful way to do such filtering (especially on generic level) is to use... intersections :) this could be a section of its own in this whole deriving types chapter. The example above can be simply rewritten as follows:
```ts
type Strings = Example & string;
```